superset-frontend/src/dashboard/components/menu/DownloadMenuItems/DownloadScreenshot.tsx

```diff
@@ -0,0 +1,149 @@
+  const anchor = currentTabs.pop();
+
+  const onDownloadScreenshot = () => {
+    let retries = 0;
+
+    // this function checks if the image is ready
+    const checkImageReady = (imageUrl: string) =>
+      fetch(`${imageUrl}?download_format=${format}`)
+        .then(response => {
+          if (response.status === 404) {
+            throw new Error('Image not ready');
+          }
+          return response.blob();
+        })
+        .then(blob => {
+          const url = window.URL.createObjectURL(blob);
+          const a = document.createElement('a');
+          a.href = url;
+          a.download = `screenshot.${format}`;
+          document.body.appendChild(a);
+          a.click();
+          document.body.removeChild(a);
+          window.URL.revokeObjectURL(url);
+        })
+        .catch(() => {
+          throw new Error('Image not ready');
+        });
+
+    // this is the functions that handles the retries
+    const fetchImageWithRetry = (imageUrl: string) => {
+      checkImageReady(imageUrl)
+        .then(() => {
+          addSuccessToast(t('The screenshot is now being downloaded.'));
+        })
+        .catch(error => {
+          // we check how many retries have been made
+          if (retries < MAX_RETRIES) {
+            retries += 1;
+            addInfoToast(
+              t(
+                'The screenshot is being generated. Please, do not leave the page.',
+              ),
+              {
+                noDuplicate: true,
+              },
+            );
+            setTimeout(() => fetchImageWithRetry(imageUrl), RETRY_INTERVAL);
+          } else {
+            addDangerToast(
+              t(
+                'The screenshot could not be downloaded. Please, try again later.',
+              ),
+            );
+            logging.error(error);
+          }
+        });
+    };
+
+    SupersetClient.post({
+      endpoint: `/api/v1/dashboard/${dashboardId}/cache_dashboard_screenshot`,
+      jsonPayload: {
+        anchor,
+      },
+    })
+      .then(({ json }) => {
+        const imageUrl = json?.image_url;
+        if (!imageUrl) {
+          throw new Error('No image URL in response');
+        }
+        addInfoToast(
+          t(
+            'The screenshot is being generated. Please, do not leave the page.',
+          ),
+        );
+        fetchImageWithRetry(imageUrl);
+      })
+      .catch(error => {
+        logging.error(error);
+        addDangerToast(
+          t('The screenshot could not be downloaded. Please, try again later.'),
+        );
+      })
+      .finally(() => {
+        logEvent?.(
+          format === DownloadScreenshotFormat.PNG
+            ? LOG_ACTIONS_DASHBOARD_DOWNLOAD_AS_IMAGE
+            : LOG_ACTIONS_DASHBOARD_DOWNLOAD_AS_PDF,
+        );
+      });
+  };
+
+  return (
+    <Menu.Item key={format} {...rest}>
+      <div onClick={onDownloadScreenshot} role="button" tabIndex={0}>
```

Comment:
I think `Menu.Item` accepts `onClick` so that div wrapper is not necessary

Reply:
This is actually necessary as the onClick is controlled at the Menu level not at the item level. At this level it would be overridden.